diff --git a/arch/riscv/Makefile b/arch/riscv/Makefile
index 972fe3492..6c585a3b9 100644
--- a/arch/riscv/Makefile
+++ b/arch/riscv/Makefile
@@ -46,6 +46,9 @@ riscv-march-aflags-$(CONFIG_RISCV_ISA_C)	:= $(riscv-march-aflags-y)c
 riscv-march-aflags-$(CONFIG_VECTOR)		:= $(riscv-march-aflags-y)v
 riscv-march-aflags-$(CONFIG_RISCV_ISA_THEAD)	:= $(riscv-march-aflags-y)xthead
 
+riscv-march-cflags-y := $(riscv-march-cflags-y)_zicsr_zifencei
+riscv-march-aflags-y := $(riscv-march-aflags-y)_zicsr_zifencei
+
 KBUILD_CFLAGS += -march=$(riscv-march-cflags-y) -Wa,-march=$(riscv-march-aflags-y)
 KBUILD_AFLAGS += -march=$(riscv-march-aflags-y)
 
diff --git a/arch/riscv/boot/dts/sunxi/Makefile b/arch/riscv/boot/dts/sunxi/Makefile
index 7a7c38bb7..8b1378917 100644
--- a/arch/riscv/boot/dts/sunxi/Makefile
+++ b/arch/riscv/boot/dts/sunxi/Makefile
@@ -1,2 +1 @@
-# SPDX-License-Identifier: GPL-2.0
-dtb-$(CONFIG_ARCH_SUNXI) += board.dtb
+
diff --git a/arch/riscv/kernel/vector.S b/arch/riscv/kernel/vector.S
index 0afc308a6..5b9abe3a8 100644
--- a/arch/riscv/kernel/vector.S
+++ b/arch/riscv/kernel/vector.S
@@ -38,14 +38,14 @@ ENTRY(__vstate_save)
 	csrr	t0,  CSR_VTYPE
 	sd	t0,  TASK_THREAD_VTYPE_V0(a0)
 
-	vsetvli	t0, x0, e8,m8
-	vsb.v	v0,  (a0)
+	.4byte  0x3072d7 /* vsetvli	t0, x0, e8,m8 */
+	.4byte  0x2050027 /* vsb.v	v0,  (a0) */
 	addi	a0, a0, RISCV_VECTOR_VLENB*8
-	vsb.v	v8,  (a0)
+	.4byte  0x2050427 /* vsb.v	v8,  (a0) */
 	addi	a0, a0, RISCV_VECTOR_VLENB*8
-	vsb.v	v16, (a0)
+	.4byte  0x2050827 /* vsb.v	v16, (a0) */
 	addi	a0, a0, RISCV_VECTOR_VLENB*8
-	vsb.v	v24, (a0)
+	.4byte  0x2050c27 /* vsb.v	v24, (a0) */
 
 	csrc	sstatus, t1
 	ret
@@ -59,14 +59,15 @@ ENTRY(__vstate_restore)
 	li	t1, (SR_VS | SR_FS)
 	csrs	sstatus, t1
 
+	.4byte  0x3072d7 /* vsetvli	t0, x0, e8,m8 */
 	vsetvli	t0, x0, e8,m8
-	vlb.v	v0,  (a0)
+	.4byte 0x12050007 /* vlb.v	v0,  (a0) */
 	addi	a0, a0, RISCV_VECTOR_VLENB*8
-	vlb.v	v8,  (a0)
+	.4byte 0x12050407 /* vlb.v	v8,  (a0) */
 	addi	a0, a0, RISCV_VECTOR_VLENB*8
-	vlb.v	v16, (a0)
+	.4byte 0x12050807 /* vlb.v	v16, (a0) */
 	addi	a0, a0, RISCV_VECTOR_VLENB*8
-	vlb.v	v24, (a0)
+	.4byte 0x12050c07 /* vlb.v	v24, (a0) */
 
 	mv	a0,  t2
 	ld	t0,  TASK_THREAD_VSTART_V0(a0)
